src/cpu-kernels/operations.cpp

```diff
@@ -116,6 +116,32 @@ ERROR awkward_listarray64_flatten_64(int64_t* tocarry, const int64_t* fromstarts
   return awkward_listarray_flatten<int64_t, int64_t>(tocarry, fromstarts, fromstops, lenstarts, startsoffset, stopsoffset);
 }
 
+template <typename C, typename T>
+ERROR awkward_listarray_transform_starts_stops(C* tostarts, C* tostops, const T* scale, const C* fromstarts, const C* fromstops,  int64_t lenstarts, int64_t startsoffset, int64_t stopsoffset, int64_t lenscale) {
+  for (int64_t i = 0; i < lenstarts; i++) {
+    int64_t start = (C)fromstarts[startsoffset + i];
+    int64_t stop = (C)fromstops[stopsoffset + i];
+    if (start < 0  ||  stop < 0) {
+      return failure("all start and stop values must be non-negative", kSliceNone, i);
+    }
+    if (lenscale != lenstarts) {
+      return failure("scale index length must be equal to start and stop lenght", kSliceNone, i);
+    }
+    tostarts[i] = start * scale[i];
+    tostops[i] = stop * scale[i];
+  }
+  return success();
+}
```

Comment:
One find-and-replace change: the name `awkward_listarray_transform_starts_stops` doesn't say it's for `flatten`, and I think that's its primary use for the foreseeable future. You can make the name less generic (and shorter, though that isn't an important goal): `awkward_listarray_flatten_scale` links it to the `flatten` operation, scaling is a kind of transformation, and `starts_stops` are implicit in the fact that this is for `listarray`.

I wouldn't have stopped this PR for a name change in cpu-kernels, but at the same time, you want to get

```c++
    if (lenscale != lenstarts) {
      return failure("scale index length must be equal to start and stop lenght", kSliceNone, i);
    }
```

out of the critical loop. (And `"lenght"` is misspelled.) You probably don't even need to pass both of them in, as one may be determined by the other in the calling code. (We do rely on correctness of the calling code; the only consistency checks we should have in cpu-kernels are the ones that depend on the values hidden in the arrays, which the calling code doesn't inspect for performance reasons.)

Finally, I haven't been able to wrap my head around how _multiplying_ the starts and stops by unequal `scale[i]` values is the right thing to do, but if the tests are sufficiently general, then I'll be convinced.

Reply:
done